Report tree-sitter errors flagged as MISSING nodes.

mjambon opened the pull request:
Update: pausing for now since it's taking longer than anticipated. I'll get back to it when I'm done with my urgent tasks.

This is a change in the ocaml-tree-sitter-core runtime library. It will surface up parsing errors that were previously unreported across all tree-sitter-based parsers used by semgrep.

This uses https://github.com/returntocorp/semgrep-rules/pull/2993. See discussion there.

test plan: usual CI tests

I'm adding a changelog entry since this affects users.

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

Commit:
Report tree-sitter errors flagged as MISSING nodes.
This is a change in the ocaml-tree-sitter-core runtime library.
It will surface up parsing errors that were previously unreported
across all tree-sitter-based parsers used by semgrep.

Commit:
Update changelog

Commit:
Mark failing test as to-do

Commit:
Update ocaml-tree-sitter-core
This fixes highlighting of empty ranges in ocaml-tree-sitter tools. This
may not be visible by semgrep users.

Commit:
Fix the syntax in some tests, move other tests to a todo/ folder

aryx commented:
This might have an impact on semgrep users, with new parsing errors coming up for people, so we want to be careful before merging it and warn people.

Commit:
Fix syntax of some tests

Commit:
Use semgrep-rules with fixes for missing nodes

mjambon commented:
I fixed the problems in semgrep-rules. In the semgrep repo, I moved away some broken tests into a `todo/` folder when I didn't know what to do.

What we need now is:

1. make sure broken patterns are reported as warnings and not as fatal errors (which is the case right now, it seems).
2. report or track the tests that were moved to `todo/` folders.

Commit:
Add JS bindings for ts_node_is_missing

Commit:
Fix syntax in test target

Commit:
Add missing semicolon in test target (Rust)

Commit:
Add missing semicolons in test targets

aryx commented:
Do I still need to review this?

mjambon commented:
@aryx I removed you because you were on vacation but then I paused the effort for 2 weeks. I'll ping the team when I need a review. I still have to check the situation of error reporting for broken patterns. I fixed a bunch broken patterns in semgrep-rules, so I'm a bit worried about the consequences on people's custom rules where they forgot a semicolon in a pattern (which is now an error).

mjambon commented:
I also have to look at CI failures.

Commit:
Try some debug printing

Commit:
Avoid tree-sitter parsing error in JS (turbo mode) by trying the pfff
parser first on C++ patterns like it's done in semgrep-core.
TODO: share the logic to avoid future discrepancies

Commit:
Update semgrep-rules

Commit:
Add todo note

Commit:
Improve error message for -dump_ast

Commit:
Add missing mli

Commit:
Remove commented-out code

Commit:
Merge branch 'develop' into mj-missing-nodes

Commit:
Revert semgrep-rules to the same version as used on develop so as to
avoid a failed test (to be fixed later)

Commit:
Fix jsoo compilation

Commit:
Add comment about not passing a literal `"<file>"` where a real file is expected.

nmote commented:
Feel free to re-add me when this is ready for review

mjambon commented:
There's a serious performance regression that I wasn't expecting at all. I can reproduce it in a striking manner: the same test (one pattern on one big js file) went from 3 s to 90 s.

mjambon commented:
For a language like JavaScript where we see the perf regression, the issue that we try two parsers one after the other until one succeeds. Since we introduced a new error condition, there will be more retrying on average. In this case, it's catastrophic because the tree-sitter parser happens to be about 30 times faster than our fallback parser and the target file is big. I think the solution is to not treat inserted (aka "missing") tokens as errors causing us to retry parsing.

Commit:
Don't fail/retry parsing with another parser if the only error is a
missing node. This uses an updated version of ocaml-tree-sitter-core
that distinguishes the two kinds of errors cleanly.

Commit:
Merge remote-tracking branch 'origin/develop' into mj-missing-nodes

Commit:
Fix bad merge

aryx commented on cli/tests/qa/test_public_repos.py:
```diff
@@ -67,6 +67,7 @@ def assert_sentinel_results(repo_path, sentinel_path, language):
     ]
 
     # nosemgrep: python.lang.security.audit.dangerous-subprocess-use.dangerous-subprocess-use
+    print(f"semgrep command: {cmd}")
```
was this added on purpose?

mjambon replied:
see other comment thread

aryx commented on cli/tests/qa/test_public_repos.py:
```diff
@@ -157,7 +158,7 @@ def test_semgrep_on_repo(monkeypatch, tmp_path, repo):
         with sentinel_path.open("w") as sentinel_file:
             sentinel_file.write(sentinel_info["file_contents"])
-
+        print(f"assert_sentinel_results({repo_path}, {sentinel_path}, {language})")
```
same? Looks like debugging statements

mjambon replied:
I forgot to remove them, but also I think all tests should print important debugging information by default.

mjambon replied:
I removed this one and kept the other one + a comment.

aryx commented on js/languages/cpp/Parser.ml:
```diff
@@ -7,7 +7,10 @@ let parse_pattern print_errors _ str =
     |> Pfff_or_tree_sitter.run_pattern ~print_errors
-         [ TreeSitterPat Parse_cpp_tree_sitter.parse_pattern ]
+         [
+           PfffPat (Parse_cpp.any_of_string Flag_parsing_cpp.Cplusplus);
```
Is this related to missing nodes?

mjambon replied:
Weird. I'll look into this.

mjambon replied:
This is the JavaScript implementation. It now does the same as the OCaml implementation. I left a comment about this coupling in the other file (`src/parsing_languages/Parse_pattern2.ml`). I'll do the same here.

mjambon replied:
Oh yeah, I didn't leave a "coupling:" comment here because it would make sense to leave one for each language (`js/languages/*/Parser.ml`).

mjambon replied:
ok, it's not that bad because only a few parsers use a non-tree-sitter parser. I added a bunch of comments.

aryx commented on src/parsing/Pfff_or_tree_sitter.ml:
```diff
@@ -36,10 +36,12 @@ type 'ast parser =
   | Pfff of (Common.filename -> 'ast * Parsing_stat.t)
   | TreeSitter of (Common.filename -> 'ast Tree_sitter_run.Parsing_result.t)
 
+(* This type is parametrized by the AST type because ??? *)
```
because we don't always generate directly generic ASTs; we sometimes generate intermediate ASTs hence the 
need for polymorphic type (so you can have Ast_cpp.program internal_result, or Ast_php.program internal_result.

aryx commented on src/parsing/Pfff_or_tree_sitter.ml:
```diff
@@ -96,14 +89,28 @@ let extract_pattern_from_tree_sitter_result
+  | Internal
+  | Error_node ->
+      true
+  | Missing_node -> false
```
maybe you could add a comment that if we consider missing_node a serious_error, this will trigger
the fallback to pfff parsers for many more cases resulting in big perf and parsing regressions, hence
this setting.

aryx reviewed (commented):
Nice work.
Why the many new parsing/todo/??? You were not familiar with the language so
just move it to todo/ because you didn't know the parsing fix?

mjambon commented:
> Why the many new parsing/todo/??? You were not familiar with the language so
> just move it to todo/ because you didn't know the parsing fix?

These should be cases where there was a parsing bug rather than a syntax error in the code. These are mostly patterns. Let me reassess and at least file bug reports.
